mreg/authentication.py

```diff
@@ -23,4 +23,8 @@ def authenticate_credentials(self, key):
         if token.created < timezone.now() - timedelta(hours=EXPIRE_HOURS):
             raise exceptions.AuthenticationFailed('Token has expired')
 
+        # Update the created time of the token to keep it valid.
+        token.created = timezone.now()
```

Comment:
If the token expiry check above used `token.updated` and that same field was updated here, you would be able to detect long-lived tokens.

Reply:
I'd say this is a good idea. Maybe use the `token.updated` field for extending the lifetime and keep `token.created` to actually be the time when it was created.

Reply:
I like this idea too.  One complication is that the `Token` model needs to be overridden.

I tried this approach:

```
diff --git a/mreg/api/views.py b/mreg/api/views.py
index 3834e62..adacb03 100644
--- a/mreg/api/views.py
+++ b/mreg/api/views.py
@@ -1,7 +1,6 @@
 from django.utils import timezone
 
 from rest_framework import status
-from rest_framework.authtoken.models import Token
 from rest_framework.authtoken.views import ObtainAuthToken
 from rest_framework.permissions import IsAuthenticated
 from rest_framework.response import Response
@@ -9,6 +8,8 @@ from rest_framework.views import APIView
 from rest_framework import serializers
 from rest_framework.exceptions import AuthenticationFailed
 
+from mreg.models import ExpiringToken
+
 
 class ObtainExpiringAuthToken(ObtainAuthToken):
 
@@ -26,9 +27,9 @@ class ObtainExpiringAuthToken(ObtainAuthToken):
         user = serializer.validated_data['user']
 
         # Force token rotation.
-        Token.objects.filter(user=user).delete()
+        ExpiringToken.objects.filter(user=user).delete()
 
-        token, created = Token.objects.get_or_create(user=user)
+        token, created = ExpiringToken.objects.get_or_create(user=user)
 
         return Response({'token': token.key})
 
diff --git a/mreg/authentication.py b/mreg/authentication.py
index cb97401..686d69f 100644
--- a/mreg/authentication.py
+++ b/mreg/authentication.py
@@ -7,24 +7,26 @@ from django.utils import timezone
 from rest_framework import exceptions
 from rest_framework.authentication import TokenAuthentication
 
+from mreg.models import ExpiringToken
+
 EXPIRE_HOURS = getattr(settings, 'REST_FRAMEWORK_TOKEN_EXPIRE_HOURS', 8)
 
 
 class ExpiringTokenAuthentication(TokenAuthentication):
     def authenticate_credentials(self, key):
         try:
-            token = self.get_model().objects.get(key=key)
+            token = ExpiringToken.objects.get(key=key)
         except ObjectDoesNotExist:
             raise exceptions.AuthenticationFailed('Invalid token')
 
         if not token.user.is_active:
             raise exceptions.AuthenticationFailed('User inactive or deleted')
 
-        if token.created < timezone.now() - timedelta(hours=EXPIRE_HOURS):
+        if token.updated < timezone.now() - timedelta(hours=EXPIRE_HOURS):
             raise exceptions.AuthenticationFailed('Token has expired')
 
         # Update the created time of the token to keep it valid.
-        token.created = timezone.now()
+        token.updated = timezone.now()
         token.save()
 
         return token.user, token
diff --git a/mreg/models.py b/mreg/models.py
index 1888364..26babd9 100644
--- a/mreg/models.py
+++ b/mreg/models.py
@@ -10,6 +10,7 @@ from django.db import DatabaseError, models, transaction
 from django.db.models import Q
 from django.utils import timezone
 from netfields import CidrAddressField, NetManager
+from rest_framework.authtoken.models import Token
 
 from .fields import DnsNameField, LCICharField
 from .models_auth import User  # noqa: F401, needed by mreg.settings for now
@@ -831,3 +832,6 @@ class BACnetID(models.Model):
             else:
                 return j
         return j
+
+class ExpiringToken(Token):
+    updated = models.DateTimeField(auto_now_add=True)

```

...which I *believe* is the correct approach (improvements welcome!); however we also have to override all the methods of [`rest_framework.authentication.TokenAuthentication`](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151).

Can you think of an easier way to solve this problem?  At this level it's tempting to just load [`knox`](https://github.com/James1345/django-rest-knox/blob/master/knox/auth.py#L25) instead.

Reply:
I think you are almost there. From:
https://github.com/encode/django-rest-framework/blob/61e7a993bd0702d30e3049179000bc7c5f284781/rest_framework/authentication.py#L164-L168
it seems easy to just set `self.model = ExpiringToken`. Probably easiest via a `__init__` method.

Reply:
Also would be nice with some tests.

Reply:
For some reason I expected `updated` to be part of the `Token` model already – but it's just a common thing to include in custom abstract base classes. Maybe use the opportunity to name it something else, like `last_used`, `last_seen` or similar.

Reply:
I also assumed `updated` was an existing field when you mentioned it, but `last_used` sounds OK.

Reply:
> I think you are almost there. From:
> 
> https://github.com/encode/django-rest-framework/blob/61e7a993bd0702d30e3049179000bc7c5f284781/rest_framework/authentication.py#L164-L168
> 
> it seems easy to just set `self.model = ExpiringToken`. Probably easiest via a `__init__` method.

Great catch, thank you!  That worked like a charm.  I've pushed a new commit that introduces `ExpiringToken`.